code/modules/shuttle/shuttle_weapons/weapon_controller.dm

```diff
@@ -0,0 +1,319 @@
+			continue
+		//Ignore ships that are on different z-levels#
+		var/obj/target_port = SSshuttle.getShuttle(ship_id)
+		if(!target_port || target_port.z != connected_port.z)
+			continue
+		if(!ship.combat_allowed)
+			continue
+		var/datum/faction/their_faction = ship.ship_faction
+		var/list/other_ship = list(
+			id = ship_id,
+			name = ship.ship_name,
+			faction = ship.ship_faction,
+			health = ship.integrity_remaining,
+			maxHealth = ship.max_ship_integrity * SHIP_INTEGRITY_FACTOR,
+			critical = ship.critical,
+			//If they consider us hostile (AI works on if we consider them hostile policy, but that is confusing for players to be shot at by 'friendly' ships).
+			hostile = check_faction_alignment(ship.ship_faction, our_ship.ship_faction) == FACTION_STATUS_HOSTILE || (their_faction.type in our_ship.rogue_factions),
+		)
+		data["ships"] += list(other_ship)
+	var/list/turfs = connected_port.return_turfs()
+	//Weapons
+	for(var/turf/T as() in turfs)
+		for(var/obj/machinery/shuttle_weapon/weapon in T)
+			var/list/active_weapon = list(
+				id = weapon.unique_id,
+				name = weapon.name,
+				cooldownLeft = max(weapon.next_shot_world_time - world.time, 0),
+				cooldown = weapon.cooldown,
+				inaccuracy = weapon.innaccuracy,
+			)
+			data["weapons"] += list(active_weapon)
+	return data
+
+/obj/machinery/computer/weapons/ui_static_data(mob/user)
+	var/list/data = list()
+	data["mapRef"] = map_name
+	return data
+
+/obj/machinery/computer/weapons/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("target_ship")
+			var/s_id = params["id"]
+			playsound(src, get_sfx("terminal_type"), 25, FALSE)
+			log_shuttle("Weapons console linked to [shuttle_id] used by [usr] set camera view to ship [s_id]")
+
+			if(!(s_id in SSbluespace_exploration.tracked_ships))
+				show_camera_static()
+				return TRUE
+
+			var/obj/docking_port/mobile/target = SSshuttle.getShuttle(s_id)
+			selected_ship_id = s_id
+
+			if(!target)
+				show_camera_static()
+				return TRUE
+
+			//Target.return_turfs() but with added range
+			var/list/L = target.return_coords()
+			var/left = min(L[1], L[3])
+			var/right = max(L[1], L[3])
+			var/top = max(L[2], L[4])
+			var/bottom = min(L[2], L[4])
+			var/turf/T0 = locate(CLAMP(left-extra_range, 1, world.maxx), CLAMP(top+extra_range, 1, world.maxy), target.z)
+			var/turf/T1 = locate(CLAMP(right+extra_range, 1, world.maxx), CLAMP(bottom-extra_range, 1, world.maxy), target.z)
+			var/list/visible_turfs = block(T0,T1)
+
+			//Corner turfs for calculations when screen is clicked.
+			//Idk why I have to subtract extra range but I do
+			corner_x = left - extra_range
+			corner_y = bottom - extra_range
+			corner_z = target.z
+
+			cam_screen.vis_contents = visible_turfs
+			cam_background.icon_state = "clear"
+
+			var/list/bbox = get_bbox_of_atoms(visible_turfs)
+			var/size_x = bbox[3] - bbox[1] + 1
+			var/size_y = bbox[4] - bbox[2] + 1
+			cam_background.fill_rect(1, 1, size_x, size_y)
```

Comment:
Unless you are doing something very complex to the `visible_turfs` list, you probably can abridge this to something like this:

```suggestion
			cam_background.fill_rect(1, 1, \
								clamp(right+extra_range, 1, world.maxx) - clamp(left-extra_range, 1, world.maxx) + 1, \
								clamp(top+extra_range, 1, world.maxy) - clamp(bottom-extra_range, 1, world.maxy) + 1)
```